proposals/threads/Overview.md

```diff
@@ -1 +1,480 @@
-TODO
+# Threading proposal for WebAssembly
+
+This page describes a proposal for the post-MVP
+[threads feature :atom_symbol:][future threads].
+
+This proposal adds a new shared linear memory type and some new operations for
+atomic memory access. The responsibility of creating and joining threads is
+deferred to the embedder.
+
+## Agents
+
+An *agent* is the execution context for a WebAssembly module. It comprises a
+module, a value stack, a control flow stack, a call stack, and an executing
+thread.
+
+The agent's executing thread evaluates instructions and modifies the value
+stack, call stack, and control flow stack as specified [here][execution spec].
+
+An agent is sometimes called a *thread*, as it is meant to match the behavior
+of the general computing concept of [threads][].
+
+## Agent Clusters
+
+An *agent cluster* is a maximal set of agents that can communicate by operating
+on shared memory.
+
+Every agent belongs to exactly one agent cluster.
+
+The embedder may deactivate or activate an agent without the agent's knowledge
+or cooperation, but must not leave some agents in the cluster active while
+other agents in the cluster are deactivated indefinitely.
+
+An embedder may terminate an agent without any of the agent's cluster's other
+agents' prior knowledge or cooperation.
+
+## Shared Linear Memory
+
+A linear memory can be marked as shared, which allows it to be shared between
+all agents in an agent cluster. The shared memory can be imported or defined in
+the module. It is a validation error to attempt to import shared linear memory
+if the module's memory import doesn't specify that it allows shared memory.
+Similarly, it is a validation error to import non-shared memory if the
+module's memory import specifies it as being shared.
+
+When memory is shared between agents, modifications to the linear memory by one
+agent can be observed by another agent in the agent cluster.
+
+### Resizing
+
+If linear memory is marked as shared, the maximum memory size must be
+specified.
+
+`grow_memory` and `current_memory` have [sequentially consistent][] ordering.
+
+`grow_memory` of a shared linear memory is allowed, but may fail if the new
+size is greater than the specified maximum size, or if reserving the additional
+memory fails. All agents in the executing agent's cluster will have access to
```

Comment:
This wording makes it a little unclear that `grow_memory` always fails if the new size is greater than the specified maximum size, which is what I assume is intended here. Possible wording suggestion:

`grow_memory` of a shared linear memory is allowed. Like non-shared memory, it fails if the new size is greater than the memory's maximum size, and it may also fail if reserving the additional memory fails.